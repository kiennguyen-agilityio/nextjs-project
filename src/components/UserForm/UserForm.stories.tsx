import type { Meta, StoryObj } from '@storybook/react';

// components
import UserForm from '@/components/UserForm';

const meta = {
  title: 'Components/UserForm',
  component: UserForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof UserForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
