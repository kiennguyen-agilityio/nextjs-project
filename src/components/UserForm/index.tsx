'use client';
import { ChangeEvent, useCallback, useState } from 'react';
import { useFormState } from 'react-dom';
import Image from 'next/image';

// models
import { UserModel } from '@/models/UserModel';

// components
import { Input } from '@/components/common/Input';
import { Button } from '@/components/common/Button';
import Dropdown from '@/components/common/Dropdown';
import { SubmitButton } from '@/components/SubmitButton';

// models

// utils
import { formatDate } from '@/utils/formatDate';

// Types
import { SelectType } from '@/types/SelectType';

// services
import { UserState, validateUser } from '@/services/validateUser';

// apis
import { uploadImage } from '@/api/image';

// constants
import { REGEX } from '@/constants/regex';
import { MAX_SIZE } from '@/constants/sizeImg';

interface UserFormProps {
  id: string;
  user?: UserModel;
  roleName: string;
  roleOptions: SelectType[];
  selectedRole: string;
  userRoleId: string;
}

const UserForm = ({
  id,
  user,
  roleName,
  roleOptions,
  selectedRole,
  userRoleId,
}: UserFormProps) => {
  const [welcomeMessage, setWelcomeMessage] = useState(
    "Welcome aboard! We are excited you are here, and we look forward to working with you. We know with your skills and experience you're a great asset to our department. If you have any questions during your first week, please contact me at any time.",
  );

  const [initialFormData] = useState({
    name: user?.name || '',
    email: user?.email || '',
    joined: user?.joined ? formatDate(user.joined) : '',
    userRole: selectedRole || '',
    avatar: user?.avatar || '',
  });

  const [formData, setFormData] = useState(initialFormData);

  const handleInputChange = (name: string, value: string) => {
    if (name === 'welcomeMessage') {
      setWelcomeMessage(value);
    } else {
      setFormData((prevState) => ({
        ...prevState,
        [name]: value,
      }));
    }
  };

  const handleRoleChange = (role: string) => {
    setFormData((prevState) => ({
      ...prevState,
      userRole: role,
    }));
  };

  const joinedDate = formData.joined ? new Date(formData.joined) : new Date(0);

  const formattedDate = formatDate(joinedDate);
  const initialState: UserState = { message: null, errors: {} };

  const updateUSerWithId = validateUser.bind(null, id);

  const [state, dispatch] = useFormState(updateUSerWithId, initialState);

  const [_, setPreviewURL] = useState<string>('');

  const handleChangeFile = useCallback(
    (callback: (value: string) => void) =>
      async (e: ChangeEvent<HTMLInputElement>) => {
        const file = (e.target.files && e.target.files[0]) as File;

        if (!file) {
          return;
        }

        if (!REGEX.IMG.test(file.name)) {
          return Error('Invalid');
        }

        // Check size of image
        if (file.size > MAX_SIZE) {
          return Error('File size is too large');
        }

        try {
          const previewImage: string = URL.createObjectURL(file);
          const formData = new FormData();

          formData.append('image', file);
          setPreviewURL(previewImage);

          const result = await uploadImage(formData);
          callback(result);
        } catch (error) {
          throw Error();
        }
      },
    [],
  );

  const hasFormChanged = () => {
    const data = JSON.stringify(formData) !== JSON.stringify(initialFormData);
    return data;
  };

  return (
    <form
      className="max-w-md mx-auto p-4 bg-white shadow-md rounded-lg"
      action={dispatch}
    >
      <h2 className="text-xl font-bold mb-4">{roleName} role</h2>
      <p className="text-sm text-gray-600 mb-4">
        Everyone who works on your {roleName} can have different roles
        <a href="#" className="text-blue-500 ml-1">
          learn more
        </a>
      </p>
      <div className="mb-4">
        <Input
          type="text"
          value={formData.name}
          customClass="mt-1 p-2 w-full border rounded-md"
          name="name"
          onChange={(value) => handleInputChange('name', value)}
          label="Type Candidate Name"
          aria-describedby="name-error"
        />
        <div id="name-error" aria-live="polite" aria-atomic="true">
          {state.errors?.name &&
            state.errors.name.map((error: string) => (
              <p className="mt-2 text-sm text-red-600" key={error}>
                {error}
              </p>
            ))}
        </div>
      </div>

      <div className="mb-4">
        <Input
          type="email"
          value={formData.email}
          customClass="mt-1 p-2 w-full border rounded-md"
          name="email"
          onChange={(value) => handleInputChange('email', value)}
          label="Type Candidate Email"
          aria-describedby="email-error"
        />
        <div id="name-error" aria-live="polite" aria-atomic="true">
          {state.errors?.email &&
            state.errors.email.map((error: string) => (
              <p className="mt-2 text-sm text-red-600" key={error}>
                {error}
              </p>
            ))}
        </div>
      </div>

      <div className="mb-4">
        <Input
          type="date"
          value={formattedDate}
          customClass="mt-1 p-2 w-full border rounded-md"
          name="joined"
          onChange={(value) => handleInputChange('joined', value)}
          label="Join Date"
          aria-describedby="joined-error"
        />
        <div id="name-error" aria-live="polite" aria-atomic="true">
          {state.errors?.joined &&
            state.errors.joined.map((error: string) => (
              <p className="mt-2 text-sm text-fill-danger" key={error}>
                {error}
              </p>
            ))}
        </div>
      </div>

      <div className="mb-4">
        <label className="block text-sm font-medium text-gray-700 mb-1">
          Avatar
        </label>
        <div className="flex items-center space-x-4">
          <div className="relative w-1/2">
            {formData.avatar ? (
              <Image
                src={formData.avatar}
                alt="Current Avatar"
                className="w-full h-28 object-cover rounded-md"
                width={100}
                height={100}
              />
            ) : (
              <p className="text-gray-500">No file chosen</p>
            )}
            <input
              type="file"
              className="absolute inset-0 opacity-0 cursor-pointer"
              name="avatar"
              onChange={handleChangeFile((value) =>
                handleInputChange('avatar', value),
              )}
              aria-describedby="avatar-error"
            />
          </div>
        </div>

        <div id="avatar-error" aria-live="polite" aria-atomic="true">
          {state.errors?.avatar &&
            state.errors.avatar.map((error: string) => (
              <p className="mt-2 text-sm text-red-600" key={error}>
                {error}
              </p>
            ))}
        </div>
      </div>

      <div className="mb-4">
        <label className="block text-sm font-medium text-gray-700">
          Select Role
        </label>
        <Dropdown
          label={selectedRole}
          options={roleOptions}
          onChange={handleRoleChange}
          value={userRoleId}
          customClass="mt-1 p-2 min-w-full rounded-md border py-4"
          aria-describedby="userRole-error"
        />
        <div id="name-error" aria-live="polite" aria-atomic="true">
          {state.errors?.userRole &&
            state.errors.userRole.map((error: string) => (
              <p className="mt-2 text-sm text-red-600" key={error}>
                {error}
              </p>
            ))}
        </div>
      </div>

      <div className="mb-4">
        <label className="block text-sm font-medium text-gray-700">
          Welcome Message
        </label>
        <textarea
          className="mt-1 p-2 w-full border rounded-md"
          rows={4}
          name="welcomeMessage"
          value={welcomeMessage}
          onChange={(e) => handleInputChange('welcomeMessage', e.target.value)}
        />
      </div>
      <div className="flex justify-end space-x-2">
        <Button customClass="px-4 py-2 border rounded-md" variant="outline">
          Cancel
        </Button>
        <SubmitButton label="Update User" disabled={!hasFormChanged()} />
      </div>
    </form>
  );
};

export default UserForm;
