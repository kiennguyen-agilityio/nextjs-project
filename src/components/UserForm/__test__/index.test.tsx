import { render } from '@testing-library/react';

// components
import UserForm from '@/components/UserForm';

test('renders user form', () => {
  const { asFragment } = render(<UserForm />);
  expect(asFragment()).toMatchSnapshot();
});
