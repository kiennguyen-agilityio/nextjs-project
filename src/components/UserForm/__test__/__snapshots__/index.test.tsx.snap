// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm renders UserForm correctly 1`] = `
<DocumentFragment>
  <form
    class="max-w-md mx-auto p-4 bg-white shadow-md rounded-lg"
  >
    <h2
      class="text-xl font-bold mb-4"
    >
      Admin role
    </h2>
    <p
      class="text-sm text-gray-600 mb-4"
    >
      Everyone who works on your Admin can have different roles
      <a
        class="text-blue-500 ml-1"
        href="#"
      >
        learn more
      </a>
    </p>
    <div
      class="mb-4"
    >
      <div>
        <label
          class="block text-dark-100 text-sm font-medium leading-5 mb-4"
          for="name"
        >
          Type Candidate Name
        </label>
        <input
          class="rounded-[10px] w-full h-12 py-2 px-4 bg-white-300 text-dark-90 placeholder-dark-50 leading-tight border border-gray-300 focus:outline-none focus:shadow-outline focus:ring-2 focus:ring-slate-300 mt-1 p-2 w-full border rounded-md"
          id="name"
          name="name"
          type="text"
          value="name 1"
        />
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        id="name-error"
      />
    </div>
    <div
      class="mb-4"
    >
      <div>
        <label
          class="block text-dark-100 text-sm font-medium leading-5 mb-4"
          for="email"
        >
          Type Candidate Email
        </label>
        <input
          class="rounded-[10px] w-full h-12 py-2 px-4 bg-white-300 text-dark-90 placeholder-dark-50 leading-tight border border-gray-300 focus:outline-none focus:shadow-outline focus:ring-2 focus:ring-slate-300 mt-1 p-2 w-full border rounded-md"
          id="email"
          name="email"
          type="email"
          value="email 1"
        />
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        id="name-error"
      />
    </div>
    <div
      class="mb-4"
    >
      <div>
        <label
          class="block text-dark-100 text-sm font-medium leading-5 mb-4"
          for="joined"
        >
          Join Date
        </label>
        <input
          class="rounded-[10px] w-full h-12 py-2 px-4 bg-white-300 text-dark-90 placeholder-dark-50 leading-tight border border-gray-300 focus:outline-none focus:shadow-outline focus:ring-2 focus:ring-slate-300 mt-1 p-2 w-full border rounded-md"
          id="joined"
          name="joined"
          type="date"
          value="2022-01-01"
        />
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        id="name-error"
      />
    </div>
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-gray-700 mb-1"
      >
        Avatar
      </label>
      <div
        class="flex items-center space-x-4"
      >
        <div
          class="relative w-1/2"
        >
          <img
            alt="Current Avatar"
            class="w-full h-28 object-cover rounded-md"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fs.net.vn%2FUr2Q&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fs.net.vn%2FUr2Q&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fs.net.vn%2FUr2Q&w=256&q=75 2x"
            style="color: transparent;"
            width="100"
          />
          <input
            aria-describedby="avatar-error"
            class="absolute inset-0 opacity-0 cursor-pointer"
            name="avatar"
            required=""
            type="file"
          />
        </div>
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        id="avatar-error"
      />
    </div>
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-gray-700"
      >
        Select Role
      </label>
      <select
        class="inline-flex justify-center w-full rounded-3xl px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-200 focus:outline-none max-w-32 mt-1 p-2 min-w-full rounded-md border py-4"
        name="userRole"
      >
        <option
          value="78e731027d8fd50ed642340b7c9a63b3"
        >
          Admin
        </option>
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
      </select>
      <div
        aria-atomic="true"
        aria-live="polite"
        id="name-error"
      />
    </div>
    <div
      class="mb-4"
    >
      <label
        class="block text-sm font-medium text-gray-700"
      >
        Welcome Message
      </label>
      <textarea
        class="mt-1 p-2 w-full border rounded-md"
        name="welcomeMessage"
        rows="4"
      >
        Welcome aboard! We are excited you are here, and we look forward to working with you. We know with your skills and experience you're a great asset to our department. If you have any questions during your first week, please contact me at any time.
      </textarea>
    </div>
    <div
      class="flex justify-end space-x-2"
    >
      <a
        href="/users"
      >
        <button
          class="flex items-center rounded p-2 border border-gray-200 dark:border-gray-700 hover:bg-gray-100 hover:opacity-80 px-4 py-2 border rounded-md"
          type="button"
        >
          Cancel
        </button>
      </a>
      <button
        class="flex items-center rounded p-2 bg-[#4270ec] text-white hover:bg-blue-600 hover:cursor-not-allowed hover:opacity-70 px-4 py-2 bg-blue-500 text-white rounded-md"
        disabled=""
        type="submit"
      >
        Update User
      </button>
    </div>
  </form>
</DocumentFragment>
`;
